indexing-service/src/main/java/org/apache/druid/indexing/firehose/WindowedSegment.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.firehose;
+
+import org.apache.druid.timeline.DataSegment;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A WindowedSegment represents a segment plus the list of intervals inside it which contribute to a timeline.
+ * If the list of intervals is null, the entire segment contributes to the timeline.
+ *
+ * This class is intended for in-memory use; WindowedSegmentId is better for serializing in specs.
+ */
+public class WindowedSegment
```

Comment:
The goal of this class looks similar to `TimelineObjectHolder`. `TimelineObjectHolder` is to represent each non-overshadowed logical segment backed by a real segment which might be partially overshadowed. `WindowedSegment` is opposite: It represents a real segment and what intervals are not overshadowed. Can we reuse `TimelineObjectHolder`?

Reply:
I think that's a good idea. I still want to have `WindowedSegmentId` (maybe with a different name) as `TimelineObjectHolder` isn't serializable, but I could reduce the delta of this PR by using `TimelineObjectHolder` instead of `WindowedSegment`.

The only trickiness is that the only code Druid has to construct `List<TimelineObjectHolder>` is `VersionedIntervalTimeline`, and [as described above](https://github.com/apache/incubator-druid/pull/7048#issuecomment-462904598) this isn't appropriate for a sub-task which only has a few segment IDs. So I have to construct it by hand, including making `PartitionChunk`s and such. But that's probably pretty straightforward.